// interface-kit/dist/core.js
var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};

// src/core/elementInfo.ts
var SVG_ELEMENTS = /* @__PURE__ */ new Set(["path", "circle", "rect", "line", "g"]);
var HEADING_ELEMENTS = /* @__PURE__ */ new Set(["h1", "h2", "h3", "h4", "h5", "h6"]);
var CONTAINER_ELEMENTS = /* @__PURE__ */ new Set([
  "div",
  "section",
  "article",
  "nav",
  "header",
  "footer",
  "aside",
  "main"
]);
var INTERACTIVE_TAGS = /* @__PURE__ */ new Set(["button", "a", "input", "select", "textarea"]);
var MODULE_HASH_PATTERN = /[A-Z0-9]{5,}/;
var CLASS_HASH_SUFFIX = /_[a-zA-Z0-9]{5,}$/;
var SHORT_CLASS_PATTERN = /^[a-z]{1,2}$/;
var SPECIAL_SELECTOR_CHARS = /[[\]():#,>+~*./]/;
var GENERIC_LAYOUT_CLASSES = /* @__PURE__ */ new Set([
  "absolute",
  "relative",
  "fixed",
  "sticky",
  "flex",
  "grid",
  "block",
  "inline",
  "hidden",
  "overflow-hidden",
  "overflow-auto",
  "overflow-scroll"
]);
var ELEMENT_IDS = /* @__PURE__ */ new WeakMap();
var elementIdCounter = 0;
function getElementId(element) {
  const existing = ELEMENT_IDS.get(element);
  if (existing) return existing;
  const next = `ik-${++elementIdCounter}`;
  ELEMENT_IDS.set(element, next);
  return next;
}
function toRectSnapshot(rect) {
  return {
    left: rect.left,
    top: rect.top,
    width: rect.width,
    height: rect.height
  };
}
function getElementPath(target, maxDepth = 4) {
  const parts = [];
  let current = target;
  let depth = 0;
  while (current && depth < maxDepth) {
    const tag = current.tagName.toLowerCase();
    if (tag === "html" || tag === "body") break;
    let identifier = tag;
    if (current.id) {
      identifier = `#${current.id}`;
    } else if (current.className && typeof current.className === "string") {
      const meaningfulClass = current.className.split(/\s+/).find(
        (className) => className.length > 2 && !SHORT_CLASS_PATTERN.test(className) && !MODULE_HASH_PATTERN.test(className)
      );
      if (meaningfulClass) {
        identifier = `.${meaningfulClass.split("_")[0]}`;
      }
    }
    parts.unshift(identifier);
    current = current.parentElement;
    depth += 1;
  }
  return parts.join(" > ");
}
function identifyElement(target) {
  var _a, _b, _c, _d, _e, _f;
  const path = getElementPath(target);
  if (target.dataset.element) {
    return { name: target.dataset.element, path };
  }
  const tag = target.tagName.toLowerCase();
  if (SVG_ELEMENTS.has(tag)) {
    const svg = target.closest("svg");
    if (svg) {
      const parent = svg.parentElement;
      if (parent) {
        const parentName = identifyElement(parent).name;
        return { name: `graphic in ${parentName}`, path };
      }
    }
    return { name: "graphic element", path };
  }
  if (tag === "svg") {
    const parent = target.parentElement;
    if ((parent == null ? void 0 : parent.tagName.toLowerCase()) === "button") {
      const buttonText = (_a = parent.textContent) == null ? void 0 : _a.trim();
      return {
        name: buttonText ? `icon in "${buttonText}" button` : "button icon",
        path
      };
    }
    return { name: "icon", path };
  }
  if (tag === "button") {
    const text = (_b = target.textContent) == null ? void 0 : _b.trim();
    const ariaLabel = target.getAttribute("aria-label");
    if (ariaLabel) return { name: `button [${ariaLabel}]`, path };
    return { name: text ? `button "${text.slice(0, 25)}"` : "button", path };
  }
  if (tag === "a") {
    const text = (_c = target.textContent) == null ? void 0 : _c.trim();
    const href = target.getAttribute("href");
    if (text) return { name: `link "${text.slice(0, 25)}"`, path };
    if (href) return { name: `link to ${href.slice(0, 30)}`, path };
    return { name: "link", path };
  }
  if (tag === "input") {
    const type = target.getAttribute("type") || "text";
    const placeholder = target.getAttribute("placeholder");
    const name = target.getAttribute("name");
    if (placeholder) return { name: `input "${placeholder}"`, path };
    if (name) return { name: `input [${name}]`, path };
    return { name: `${type} input`, path };
  }
  if (HEADING_ELEMENTS.has(tag)) {
    const text = (_d = target.textContent) == null ? void 0 : _d.trim();
    return { name: text ? `${tag} "${text.slice(0, 35)}"` : tag, path };
  }
  if (tag === "p") {
    const text = (_e = target.textContent) == null ? void 0 : _e.trim();
    if (text) {
      return {
        name: `paragraph: "${text.slice(0, 40)}${text.length > 40 ? "..." : ""}"`,
        path
      };
    }
    return { name: "paragraph", path };
  }
  if (tag === "span" || tag === "label") {
    const text = (_f = target.textContent) == null ? void 0 : _f.trim();
    if (text && text.length < 40) return { name: `"${text}"`, path };
    return { name: tag, path };
  }
  if (CONTAINER_ELEMENTS.has(tag)) {
    const className = target.className;
    const role = target.getAttribute("role");
    const ariaLabel = target.getAttribute("aria-label");
    if (ariaLabel) return { name: `${tag} [${ariaLabel}]`, path };
    if (role === "button" || looksLikeButton(target)) {
      return { name: "Button", path };
    }
    if (typeof className === "string" && className) {
      const tailwindPrefixes = /^(bg-|text-|px-|py-|pt-|pb-|pl-|pr-|p-|m-|mt-|mb-|ml-|mr-|mx-|my-|w-|h-|min-|max-|flex|grid|block|inline|hidden|absolute|relative|fixed|sticky|rounded|border|shadow|opacity|z-|gap-|space-|font-|leading-|tracking-|overflow|cursor|pointer|select|transition|duration|ease|animate|hover:|focus:|active:|dark:)/;
      const words = className.split(/\s+/).filter((value) => !tailwindPrefixes.test(value)).map((value) => value.replace(MODULE_HASH_PATTERN, "").replace(CLASS_HASH_SUFFIX, "")).filter((value) => value.length > 2 && !SHORT_CLASS_PATTERN.test(value)).slice(0, 2);
      if (words.length > 0) return { name: words.join(" "), path };
    }
    return { name: tag === "div" ? "Container" : tag, path };
  }
  return { name: tag, path };
}
function getElementClasses(target) {
  const className = target.className;
  if (typeof className !== "string" || !className) return [];
  const seen = /* @__PURE__ */ new Set();
  const result = [];
  for (const classNamePart of className.split(/\s+/)) {
    if (!classNamePart) continue;
    const cleaned = classNamePart.replace(CLASS_HASH_SUFFIX, "");
    if (!seen.has(cleaned)) {
      seen.add(cleaned);
      result.push(cleaned);
    }
  }
  return result;
}
function generateSelector(target) {
  const tag = target.tagName.toLowerCase();
  const id = target.id;
  const ownerDocument = target.ownerDocument;
  if (id && !SPECIAL_SELECTOR_CHARS.test(id)) {
    return `${tag}#${id}`;
  }
  const safeClasses = getElementClasses(target).filter(
    (className) => !SPECIAL_SELECTOR_CHARS.test(className)
  );
  let selector = tag;
  if (safeClasses.length > 0) {
    selector += `.${safeClasses.slice(0, 3).join(".")}`;
  }
  try {
    const matches = ownerDocument.querySelectorAll(selector);
    if (matches.length === 1) return selector;
  } catch (e) {
  }
  const parent = target.parentElement;
  if (parent) {
    const siblings = Array.from(parent.children);
    const index = siblings.indexOf(target) + 1;
    return `${selector}:nth-child(${index})`;
  }
  return selector;
}
function buildElementInfo(element, full) {
  var _a;
  const { name, path } = identifyElement(element);
  const classes = getElementClasses(element);
  return {
    key: getElementId(element),
    name,
    path,
    tag: element.tagName.toLowerCase(),
    id: element.id || null,
    classes,
    textContent: full ? ((_a = element.textContent) == null ? void 0 : _a.trim().slice(0, 100)) || "" : "",
    selector: full ? generateSelector(element) : "",
    similarCount: full ? findSimilarElements(element).length : 0
  };
}
function extractElementInfo(element) {
  return buildElementInfo(element, true);
}
function extractHoverInfo(element) {
  const { name, path } = identifyElement(element);
  return {
    key: getElementId(element),
    name,
    path,
    tag: element.tagName.toLowerCase(),
    rect: toRectSnapshot(element.getBoundingClientRect())
  };
}
function looksLikeButton(element) {
  var _a;
  const className = typeof element.className === "string" ? element.className : "";
  const text = ((_a = element.textContent) == null ? void 0 : _a.trim()) || "";
  const hasCursorPointer = className.includes("cursor-pointer");
  const hasShortText = text.length > 0 && text.length < 30;
  const hasRounded = className.includes("rounded");
  const hasBackground = className.includes("bg-");
  const hasOnClick = element.hasAttribute("onclick") || element.closest("[onclick]") !== null;
  const hasButtonClass = /\b(btn|button|cta)\b/i.test(className);
  if (hasCursorPointer && hasShortText && (hasRounded || hasBackground)) {
    return true;
  }
  if (hasButtonClass) {
    return true;
  }
  if (hasOnClick && hasShortText) {
    return true;
  }
  return false;
}
function isInteractiveElement(element) {
  const tag = element.tagName.toLowerCase();
  if (INTERACTIVE_TAGS.has(tag)) return true;
  if (element.getAttribute("role") === "button") return true;
  if (element.hasAttribute("onclick")) return true;
  if (element.classList.contains("btn") || element.classList.contains("button")) {
    return true;
  }
  return false;
}
function findSimilarElements(element) {
  const tag = element.tagName.toLowerCase();
  const classes = getElementClasses(element);
  if (classes.length === 0) return [];
  const meaningfulClasses = classes.filter(
    (className) => !SHORT_CLASS_PATTERN.test(className) && !SPECIAL_SELECTOR_CHARS.test(className) && className.length > 2 && !GENERIC_LAYOUT_CLASSES.has(className)
  );
  if (meaningfulClasses.length === 0) return [];
  const meaningfulSet = new Set(meaningfulClasses);
  const selectorClasses = meaningfulClasses.slice(0, 5);
  const selector = `${tag}.${selectorClasses.join(".")}`;
  const MAX_DEPTH = 4;
  const MAX_SIMILAR = 50;
  let bestMatches = [];
  let ancestor = element.parentElement;
  for (let index = 0; index < MAX_DEPTH && ancestor; index += 1) {
    try {
      const matches = Array.from(ancestor.querySelectorAll(selector));
      if (matches.length > 1 && matches.length <= MAX_SIMILAR) {
        bestMatches = matches;
      }
    } catch (e) {
    }
    ancestor = ancestor.parentElement;
  }
  if (bestMatches.length === 0) return [];
  return bestMatches.filter((candidate) => {
    if (candidate === element) return false;
    if (element.contains(candidate) || candidate.contains(element)) return false;
    const candidateMeaningful = getElementClasses(candidate).filter(
      (className) => !SHORT_CLASS_PATTERN.test(className) && !SPECIAL_SELECTOR_CHARS.test(className) && className.length > 2 && !GENERIC_LAYOUT_CLASSES.has(className)
    );
    const overlap = candidateMeaningful.filter((className) => meaningfulSet.has(className)).length;
    const smaller = Math.min(meaningfulSet.size, candidateMeaningful.length);
    if (smaller === 0 || overlap / smaller < 0.6) return false;
    return true;
  });
}
function findBestElement(element) {
  if (isInteractiveElement(element)) return element;
  let current = element;
  let depth = 0;
  while (current && depth < 3) {
    if (isInteractiveElement(current)) {
      return current;
    }
    current = current.parentElement;
    depth += 1;
  }
  return element;
}

// src/core/controller.ts
var INTERFACE_KIT_ATTR = "data-interface-kit";
var CURSOR_STYLE_ID = "interface-kit-cursor";
var TEXT_ELEMENT_SELECTOR = "span, p, h1, h2, h3, h4, h5, h6, a, label";
var TEXT_PROPERTIES = /* @__PURE__ */ new Set([
  "font-size",
  "font-weight",
  "color",
  "text-align",
  "line-height",
  "letter-spacing"
]);
var TEXT_TAGS = /* @__PURE__ */ new Set(["SPAN", "P", "H1", "H2", "H3", "H4", "H5", "H6", "A", "LABEL", "BUTTON"]);
var cursorStyleUsers = 0;
function resolveEnabled(explicit) {
  if (typeof explicit === "boolean") return explicit;
  if (typeof process !== "undefined" && process.env.NODE_ENV) {
    return process.env.NODE_ENV !== "production";
  }
  return false;
}
function ensureCursorStyle(ownerDocument) {
  cursorStyleUsers += 1;
  if (ownerDocument.getElementById(CURSOR_STYLE_ID)) return;
  const style = ownerDocument.createElement("style");
  style.id = CURSOR_STYLE_ID;
  style.textContent = `
    body { cursor: crosshair !important; }
    body * { cursor: crosshair !important; }
    [${INTERFACE_KIT_ATTR}] { cursor: default !important; }
    [${INTERFACE_KIT_ATTR}] button,
    [${INTERFACE_KIT_ATTR}] input,
    [${INTERFACE_KIT_ATTR}] select,
    [${INTERFACE_KIT_ATTR}] textarea,
    [${INTERFACE_KIT_ATTR}] label { cursor: default !important; }
    [contenteditable="true"] {
      cursor: text !important;
      caret-color: #2563eb !important;
      outline: none !important;
    }
  `;
  ownerDocument.head.appendChild(style);
}
function releaseCursorStyle(ownerDocument) {
  var _a;
  cursorStyleUsers = Math.max(0, cursorStyleUsers - 1);
  if (cursorStyleUsers > 0) return;
  (_a = ownerDocument.getElementById(CURSOR_STYLE_ID)) == null ? void 0 : _a.remove();
}
function isRectEqual(a, b) {
  return a.left === b.left && a.top === b.top && a.width === b.width && a.height === b.height;
}
function elementHasText(element) {
  var _a, _b, _c;
  if (TEXT_TAGS.has(element.tagName) && ((_a = element.textContent) == null ? void 0 : _a.trim())) return true;
  for (const node of Array.from(element.childNodes)) {
    if (node.nodeType === Node.TEXT_NODE && ((_b = node.textContent) == null ? void 0 : _b.trim())) return true;
    if (node.nodeType === Node.ELEMENT_NODE && node instanceof HTMLElement && TEXT_TAGS.has(node.tagName) && ((_c = node.textContent) == null ? void 0 : _c.trim())) {
      return true;
    }
  }
  return false;
}
function getMutationTargets(element, property) {
  if (!TEXT_PROPERTIES.has(property)) return [element];
  const targets = [element];
  element.querySelectorAll(TEXT_ELEMENT_SELECTOR).forEach((child) => {
    if (child instanceof HTMLElement) {
      targets.push(child);
    }
  });
  return targets;
}
function eventIsWithinInterfaceKit(event) {
  return event.composedPath().some(
    (node) => node instanceof HTMLElement && (node.hasAttribute(INTERFACE_KIT_ATTR) || Boolean(node.closest(`[${INTERFACE_KIT_ATTR}]`)))
  );
}
var Controller = class {
  constructor(options = {}) {
    this.listeners = /* @__PURE__ */ new Set();
    this.liveElements = /* @__PURE__ */ new Map();
    this.similarElementsCache = /* @__PURE__ */ new Map();
    this.originalValues = /* @__PURE__ */ new Map();
    this.modifiedElements = /* @__PURE__ */ new Map();
    this.pendingStyleWrites = /* @__PURE__ */ new Map();
    this.hoverFrame = null;
    this.flushFrame = null;
    this.hoverPoint = null;
    this.originalText = null;
    this.mounted = false;
    this.isCursorActive = false;
    this.flushStyleWrites = () => {
      this.flushFrame = null;
      for (const [element, writes] of this.pendingStyleWrites) {
        if (!element.isConnected) {
          this.pendingStyleWrites.delete(element);
          this.modifiedElements.delete(element);
          this.originalValues.delete(element);
          continue;
        }
        for (const [property, value] of writes) {
          if (value === null) {
            this.restoreCapturedStyle(element, property);
          } else {
            element.style.setProperty(property, value, "important");
          }
        }
      }
      this.pendingStyleWrites.clear();
    };
    this.processHover = () => {
      this.hoverFrame = null;
      if (!this.state.isActive || this.state.isEditingText || this.state.selectedElement) {
        this.setHoveredElement(null);
        return;
      }
      const hoverPoint = this.hoverPoint;
      if (!hoverPoint || !this.ownerDocument) return;
      const elementUnder = this.ownerDocument.elementFromPoint(hoverPoint.x, hoverPoint.y);
      if (!(elementUnder instanceof HTMLElement)) {
        this.setHoveredElement(null);
        return;
      }
      if (elementUnder.closest(`[${INTERFACE_KIT_ATTR}]`)) {
        this.setHoveredElement(null);
        return;
      }
      const bestElement = findBestElement(elementUnder);
      this.setHoveredElement(extractHoverInfo(bestElement));
    };
    this.handleMouseMove = (event) => {
      var _a;
      if (!this.state.isActive || this.state.isEditingText || this.state.selectedElement) {
        return;
      }
      if (eventIsWithinInterfaceKit(event)) {
        this.setHoveredElement(null);
        return;
      }
      this.hoverPoint = { x: event.clientX, y: event.clientY };
      if (this.hoverFrame !== null || !((_a = this.ownerDocument) == null ? void 0 : _a.defaultView)) return;
      this.hoverFrame = this.ownerDocument.defaultView.requestAnimationFrame(
        this.processHover
      );
    };
    this.handleClick = (event) => {
      if (!this.state.isActive || !this.ownerDocument) return;
      if (eventIsWithinInterfaceKit(event)) return;
      const elementUnder = this.ownerDocument.elementFromPoint(event.clientX, event.clientY);
      if (!(elementUnder instanceof HTMLElement)) return;
      if (elementUnder.closest(`[${INTERFACE_KIT_ATTR}]`)) return;
      if (elementUnder.closest('[contenteditable="true"]')) return;
      const activeEditable = this.ownerDocument.querySelector(
        '[contenteditable="true"]:not([data-interface-kit])'
      );
      if (activeEditable instanceof HTMLElement) {
        this.exitTextEditing(true);
        this.clearSelection();
        event.preventDefault();
        event.stopPropagation();
        return;
      }
      event.preventDefault();
      event.stopPropagation();
      const bestElement = findBestElement(elementUnder);
      const selectedElement = this.getSelectedElement();
      if (selectedElement === bestElement) {
        return;
      }
      this.setSelectedElement(extractElementInfo(bestElement), bestElement);
    };
    this.handleDoubleClick = (event) => {
      if (!this.state.isActive || !this.ownerDocument) return;
      if (eventIsWithinInterfaceKit(event)) return;
      const elementUnder = this.ownerDocument.elementFromPoint(event.clientX, event.clientY);
      if (!(elementUnder instanceof HTMLElement)) return;
      if (elementUnder.closest(`[${INTERFACE_KIT_ATTR}]`)) return;
      const bestElement = findBestElement(elementUnder);
      if (!bestElement.isConnected || !elementHasText(bestElement)) return;
      event.preventDefault();
      event.stopPropagation();
      const selectedElement = this.getSelectedElement();
      if (!this.state.selectedElement || selectedElement !== bestElement) {
        this.setSelectedElement(extractElementInfo(bestElement), bestElement);
      }
      this.enterTextEditing(bestElement);
    };
    this.handleKeyDown = (event) => {
      if (event.key !== "Escape") return;
      if (this.state.isEditingText) {
        event.preventDefault();
        this.exitTextEditing(true);
        return;
      }
      if (this.state.selectedElement) {
        this.clearSelection();
        return;
      }
      if (this.state.isActive) {
        this.setActive(false);
      }
    };
    this.options = options;
    this.ownerDocument = options.ownerDocument || (typeof document !== "undefined" ? document : void 0);
    this.enabled = resolveEnabled(options.enabled);
    this.state = {
      isMounted: false,
      isActive: Boolean(options.initialActive),
      activeTab: "style",
      selectedElement: null,
      hoveredElement: null,
      isEditingText: false,
      isDraggingStyle: false,
      pendingChanges: []
    };
  }
  mount() {
    if (this.mounted || !this.enabled || !this.ownerDocument) return;
    this.mounted = true;
    this.ownerDocument.addEventListener("mousemove", this.handleMouseMove, {
      passive: true
    });
    this.ownerDocument.addEventListener("click", this.handleClick, true);
    this.ownerDocument.addEventListener("dblclick", this.handleDoubleClick, true);
    this.ownerDocument.addEventListener("keydown", this.handleKeyDown);
    this.setSnapshot({ isMounted: true });
    this.syncCursorState();
  }
  destroy() {
    var _a, _b;
    if (!this.mounted || !this.ownerDocument) return;
    this.exitTextEditing(true);
    if (this.hoverFrame !== null) {
      (_a = this.ownerDocument.defaultView) == null ? void 0 : _a.cancelAnimationFrame(this.hoverFrame);
      this.hoverFrame = null;
    }
    if (this.flushFrame !== null) {
      (_b = this.ownerDocument.defaultView) == null ? void 0 : _b.cancelAnimationFrame(this.flushFrame);
      this.flushFrame = null;
    }
    this.ownerDocument.removeEventListener("mousemove", this.handleMouseMove);
    this.ownerDocument.removeEventListener("click", this.handleClick, true);
    this.ownerDocument.removeEventListener("dblclick", this.handleDoubleClick, true);
    this.ownerDocument.removeEventListener("keydown", this.handleKeyDown);
    this.revertAllChanges();
    this.liveElements.clear();
    this.similarElementsCache.clear();
    this.pendingStyleWrites.clear();
    if (this.isCursorActive) {
      releaseCursorStyle(this.ownerDocument);
      this.isCursorActive = false;
    }
    this.mounted = false;
    this.setSnapshot({
      isMounted: false,
      isActive: false,
      hoveredElement: null,
      selectedElement: null,
      isEditingText: false,
      isDraggingStyle: false,
      pendingChanges: []
    });
  }
  setActive(active) {
    if (!this.enabled) return;
    if (!active) {
      this.exitTextEditing(true);
      this.similarElementsCache.clear();
      this.setSnapshot({
        isActive: false,
        selectedElement: null,
        hoveredElement: null,
        isEditingText: false,
        isDraggingStyle: false
      });
    } else {
      this.setSnapshot({ isActive: true });
    }
    this.syncCursorState();
  }
  toggle() {
    this.setActive(!this.state.isActive);
  }
  clearSelection() {
    this.exitTextEditing(true);
    this.similarElementsCache.clear();
    this.setSnapshot({
      selectedElement: null,
      hoveredElement: null,
      isEditingText: false,
      isDraggingStyle: false
    });
  }
  setTab(tab) {
    if (this.state.activeTab === tab) return;
    this.setSnapshot({ activeTab: tab });
  }
  getState() {
    return this.state;
  }
  subscribe(listener) {
    this.listeners.add(listener);
    return () => {
      this.listeners.delete(listener);
    };
  }
  getSelectedElement() {
    var _a;
    return this.getElementByKey((_a = this.state.selectedElement) == null ? void 0 : _a.key);
  }
  applyStyle(element, property, value, tailwindClass, targetElementInfo) {
    this.applyStyleGroup(
      [element, ...this.getSimilarElements(element)],
      property,
      value,
      tailwindClass,
      targetElementInfo
    );
  }
  applyStyleGroup(elements, property, value, tailwindClass, targetElementInfo) {
    var _a;
    const uniqueElements = Array.from(
      new Set(elements.filter((element) => element == null ? void 0 : element.isConnected))
    );
    if (uniqueElements.length === 0) return;
    const firstElement = uniqueElements[0];
    const originalValue = this.captureOriginalStyle(firstElement, property).computedValue;
    for (const element of uniqueElements) {
      this.registerElement(element);
      for (const target of getMutationTargets(element, property)) {
        this.captureOriginalStyle(target, property);
        if (!this.modifiedElements.has(target)) {
          this.modifiedElements.set(target, /* @__PURE__ */ new Set());
        }
        (_a = this.modifiedElements.get(target)) == null ? void 0 : _a.add(property);
        this.queueStyleWrite(target, property, value);
      }
    }
    const trackAs = targetElementInfo || this.state.selectedElement;
    if (!trackAs) return;
    const existingIndex = this.state.pendingChanges.findIndex(
      (change2) => change2.elementInfo.selector === trackAs.selector && change2.property === property
    );
    const nextChanges = [...this.state.pendingChanges];
    const change = {
      id: `${trackAs.selector}-${property}`,
      elementInfo: trackAs,
      property,
      originalValue,
      newValue: value,
      tailwindClass
    };
    if (existingIndex >= 0) {
      nextChanges[existingIndex] = change;
    } else {
      nextChanges.push(change);
    }
    this.setSnapshot({ pendingChanges: nextChanges });
  }
  revertProperty(element, property, targetElementInfo) {
    this.revertPropertyGroup(
      [element, ...this.getSimilarElements(element)],
      property,
      targetElementInfo
    );
  }
  revertPropertyGroup(elements, property, targetElementInfo) {
    const uniqueElements = Array.from(
      new Set(elements.filter((element) => element == null ? void 0 : element.isConnected))
    );
    if (uniqueElements.length === 0) return;
    for (const element of uniqueElements) {
      for (const target of getMutationTargets(element, property)) {
        this.queueStyleWrite(target, property, null);
      }
    }
    const trackAs = targetElementInfo || this.state.selectedElement;
    if (!trackAs) return;
    this.setSnapshot({
      pendingChanges: this.state.pendingChanges.filter(
        (change) => !(change.elementInfo.selector === trackAs.selector && change.property === property)
      )
    });
  }
  revertAllChanges() {
    var _a, _b;
    if (this.flushFrame !== null && ((_a = this.ownerDocument) == null ? void 0 : _a.defaultView)) {
      this.ownerDocument.defaultView.cancelAnimationFrame(this.flushFrame);
      this.flushFrame = null;
    }
    this.pendingStyleWrites.clear();
    for (const change of this.state.pendingChanges) {
      if (change.property !== "text-content") continue;
      try {
        const element = (_b = this.ownerDocument) == null ? void 0 : _b.querySelector(change.elementInfo.selector);
        if (element instanceof HTMLElement) {
          element.textContent = change.originalValue;
        }
      } catch (e) {
      }
    }
    for (const [element, properties] of this.modifiedElements) {
      for (const property of properties) {
        this.restoreCapturedStyle(element, property);
      }
    }
    this.modifiedElements.clear();
    this.originalValues.clear();
    this.similarElementsCache.clear();
    this.setSnapshot({
      hoveredElement: null,
      pendingChanges: [],
      selectedElement: null,
      isEditingText: false,
      isDraggingStyle: false
    });
  }
  startStyleInteraction() {
    if (!this.state.isDraggingStyle) {
      this.setSnapshot({ isDraggingStyle: true });
    }
  }
  endStyleInteraction() {
    if (this.state.isDraggingStyle) {
      this.setSnapshot({ isDraggingStyle: false });
    }
  }
  notify() {
    for (const listener of this.listeners) {
      listener(this.state);
    }
  }
  setSnapshot(next) {
    this.state = __spreadValues(__spreadValues({}, this.state), next);
    this.notify();
  }
  syncCursorState() {
    if (!this.ownerDocument) return;
    if (this.state.isActive && !this.isCursorActive) {
      ensureCursorStyle(this.ownerDocument);
      this.isCursorActive = true;
    } else if (!this.state.isActive && this.isCursorActive) {
      releaseCursorStyle(this.ownerDocument);
      this.isCursorActive = false;
    }
  }
  registerElement(element) {
    if (!element) return null;
    const key = getElementId(element);
    this.liveElements.set(key, element);
    return key;
  }
  getElementByKey(key) {
    if (!key) return null;
    const element = this.liveElements.get(key);
    if (!element) return null;
    if (element.isConnected) return element;
    this.liveElements.delete(key);
    this.similarElementsCache.delete(key);
    return null;
  }
  getSimilarElements(element) {
    const key = this.registerElement(element);
    if (!key) return [];
    const cached = this.similarElementsCache.get(key);
    if (cached) {
      const live = cached.filter((candidate) => candidate.isConnected);
      this.similarElementsCache.set(key, live);
      return live;
    }
    const similar = findSimilarElements(element).filter((candidate) => {
      this.registerElement(candidate);
      return candidate.isConnected;
    });
    this.similarElementsCache.set(key, similar);
    return similar;
  }
  setSelectedElement(selected, element) {
    if (element) {
      this.registerElement(element);
    }
    this.similarElementsCache.clear();
    if (selected && (element == null ? void 0 : element.isConnected)) {
      const similar = findSimilarElements(element).filter((candidate) => {
        this.registerElement(candidate);
        return candidate.isConnected;
      });
      this.similarElementsCache.set(selected.key, similar);
    }
    this.setSnapshot({
      selectedElement: selected,
      hoveredElement: null,
      isEditingText: selected ? this.state.isEditingText : false,
      isDraggingStyle: false
    });
  }
  setHoveredElement(hovered) {
    if (!hovered) {
      if (this.state.hoveredElement !== null) {
        this.setSnapshot({ hoveredElement: null });
      }
      return;
    }
    if (this.state.hoveredElement && this.state.hoveredElement.key === hovered.key && this.state.hoveredElement.name === hovered.name && isRectEqual(this.state.hoveredElement.rect, hovered.rect)) {
      return;
    }
    this.setSnapshot({ hoveredElement: hovered });
  }
  captureOriginalStyle(element, property) {
    let elementOriginals = this.originalValues.get(element);
    if (!elementOriginals) {
      elementOriginals = /* @__PURE__ */ new Map();
      this.originalValues.set(element, elementOriginals);
    }
    const existing = elementOriginals.get(property);
    if (existing) return existing;
    const original = {
      computedValue: getComputedStyle(element).getPropertyValue(property),
      inlineValue: element.style.getPropertyValue(property),
      inlinePriority: element.style.getPropertyPriority(property)
    };
    elementOriginals.set(property, original);
    return original;
  }
  restoreCapturedStyle(element, property) {
    const elementOriginals = this.originalValues.get(element);
    const original = elementOriginals == null ? void 0 : elementOriginals.get(property);
    if (!original) {
      element.style.removeProperty(property);
    } else if (original.inlineValue) {
      element.style.setProperty(property, original.inlineValue, original.inlinePriority);
    } else {
      element.style.removeProperty(property);
    }
    const modified = this.modifiedElements.get(element);
    modified == null ? void 0 : modified.delete(property);
    if ((modified == null ? void 0 : modified.size) === 0) {
      this.modifiedElements.delete(element);
    }
    elementOriginals == null ? void 0 : elementOriginals.delete(property);
    if ((elementOriginals == null ? void 0 : elementOriginals.size) === 0) {
      this.originalValues.delete(element);
    }
  }
  scheduleFlush() {
    var _a;
    if (this.flushFrame !== null || !((_a = this.ownerDocument) == null ? void 0 : _a.defaultView)) return;
    this.flushFrame = this.ownerDocument.defaultView.requestAnimationFrame(
      this.flushStyleWrites
    );
  }
  queueStyleWrite(element, property, value) {
    let elementWrites = this.pendingStyleWrites.get(element);
    if (!elementWrites) {
      elementWrites = /* @__PURE__ */ new Map();
      this.pendingStyleWrites.set(element, elementWrites);
    }
    elementWrites.set(property, value);
    this.scheduleFlush();
  }
  enterTextEditing(element) {
    var _a, _b;
    this.originalText = ((_a = element.textContent) == null ? void 0 : _a.trim()) || "";
    element.contentEditable = "true";
    element.focus();
    const range = element.ownerDocument.createRange();
    range.selectNodeContents(element);
    const selection = (_b = element.ownerDocument.defaultView) == null ? void 0 : _b.getSelection();
    selection == null ? void 0 : selection.removeAllRanges();
    selection == null ? void 0 : selection.addRange(range);
    this.setSnapshot({ isEditingText: true });
  }
  exitTextEditing(commit) {
    var _a;
    const selected = this.state.selectedElement;
    const element = this.getSelectedElement();
    if (!selected || !element) {
      this.originalText = null;
      if (this.state.isEditingText) {
        this.setSnapshot({ isEditingText: false });
      }
      return;
    }
    if (element.contentEditable === "true") {
      const newText = ((_a = element.textContent) == null ? void 0 : _a.trim()) || "";
      const originalText = this.originalText;
      element.contentEditable = "inherit";
      if (commit && originalText !== null && newText !== originalText) {
        const nextChanges = [...this.state.pendingChanges];
        const existingIndex = nextChanges.findIndex(
          (change) => change.elementInfo.selector === selected.selector && change.property === "text-content"
        );
        const nextChange = {
          id: `${selected.selector}-text-content`,
          elementInfo: selected,
          property: "text-content",
          originalValue: originalText,
          newValue: newText
        };
        if (existingIndex >= 0) {
          nextChanges[existingIndex] = nextChange;
        } else {
          nextChanges.push(nextChange);
        }
        this.setSnapshot({
          pendingChanges: nextChanges,
          isEditingText: false
        });
      } else {
        this.setSnapshot({ isEditingText: false });
      }
    } else if (this.state.isEditingText) {
      this.setSnapshot({ isEditingText: false });
    }
    this.originalText = null;
  }
};
function createInterfaceKitController(options = {}) {
  return new Controller(options);
}

// src/core/promptFormatter.ts
var TEXT_PROPERTIES2 = /* @__PURE__ */ new Set([
  "font-size",
  "font-weight",
  "color",
  "text-align",
  "line-height",
  "letter-spacing"
]);
function getCurrentRoute() {
  return window.location.pathname;
}
function getAncestryPath(selector) {
  try {
    const element = document.querySelector(selector);
    if (!element) return "";
    const parts = [];
    let current = element;
    let depth = 0;
    while (current && depth < 4) {
      const tag = current.tagName.toLowerCase();
      if (tag === "body" || tag === "html") break;
      let identifier = tag;
      if (current.id) {
        identifier = `${tag}#${current.id}`;
      } else {
        const classes = Array.from(current.classList);
        const meaningful = classes.find(
          (className) => className.length > 3 && !className.startsWith("bg-") && !className.startsWith("text-") && !className.startsWith("p-") && !className.startsWith("m-") && !className.startsWith("flex") && !className.startsWith("grid")
        );
        if (meaningful) {
          identifier = `${tag}.${meaningful}`;
        }
      }
      parts.unshift(identifier);
      current = current.parentElement;
      depth += 1;
    }
    return parts.join(" > ");
  } catch (e) {
    return "";
  }
}
function cssToTailwind(property, value) {
  const num = parseFloat(value);
  switch (property) {
    case "padding": {
      const parts = value.split(/\s+/).map((part) => parseFloat(part));
      if (parts.length === 2) {
        const [y, x] = parts;
        return `py-[${y}px] px-[${x}px]`;
      }
      if (parts.length === 1) return `p-[${parts[0]}px]`;
      return `p-[${value}]`;
    }
    case "padding-top":
      return `pt-[${num}px]`;
    case "padding-right":
      return `pr-[${num}px]`;
    case "padding-bottom":
      return `pb-[${num}px]`;
    case "padding-left":
      return `pl-[${num}px]`;
    case "margin-top":
      return `mt-[${num}px]`;
    case "margin-right":
      return `mr-[${num}px]`;
    case "margin-bottom":
      return `mb-[${num}px]`;
    case "margin-left":
      return `ml-[${num}px]`;
    case "gap":
      return num % 4 === 0 ? `gap-${num / 4}` : `gap-[${num}px]`;
    case "border-radius":
      if (num >= 9999) return "rounded-full";
      if (num === 0) return "rounded-none";
      return `rounded-[${num}px]`;
    case "border-width":
      if (num === 0) return "border-0";
      if (num === 1) return "border";
      if (num === 2) return "border-2";
      if (num === 4) return "border-4";
      return `border-[${num}px]`;
    case "border-style":
      return `border-${value}`;
    case "background-color":
      return `bg-[${value}]`;
    case "border-color":
      return `border-[${value}]`;
    case "backdrop-filter":
      return value === "none" ? "backdrop-blur-0" : `backdrop-[${value}]`;
    case "box-shadow":
      return value === "none" ? "shadow-none" : "shadow";
    case "font-size":
      if (num === 12) return "text-xs";
      if (num === 14) return "text-sm";
      if (num === 16) return "text-base";
      if (num === 18) return "text-lg";
      if (num === 20) return "text-xl";
      if (num === 24) return "text-2xl";
      return `text-[${num}px]`;
    case "font-weight": {
      const weights = {
        "100": "font-thin",
        "200": "font-extralight",
        "300": "font-light",
        "400": "font-normal",
        "500": "font-medium",
        "600": "font-semibold",
        "700": "font-bold",
        "800": "font-extrabold",
        "900": "font-black"
      };
      return weights[value] || `font-[${value}]`;
    }
    case "width":
      return `w-[${num}px]`;
    case "height":
      return `h-[${num}px]`;
    case "opacity":
      return `opacity-[${Math.round(parseFloat(value) * 100)}%]`;
    case "text-align":
      return `text-${value}`;
    case "line-height":
      return `leading-[${value}]`;
    case "letter-spacing":
      return `tracking-[${value}]`;
    case "animation":
      return value === "none" ? "animate-none" : `animate-[${value}]`;
    default:
      return value;
  }
}
function groupChanges(changes) {
  var _a;
  const groups = /* @__PURE__ */ new Map();
  for (const change of changes) {
    const key = change.elementInfo.selector;
    if (!groups.has(key)) {
      groups.set(key, {
        selector: change.elementInfo.selector,
        tag: change.elementInfo.tag,
        text: ((_a = change.elementInfo.textContent) == null ? void 0 : _a.slice(0, 30)) || "",
        classes: change.elementInfo.classes.slice(0, 8),
        path: change.elementInfo.path,
        changes: [],
        childChanges: []
      });
    }
    const group = groups.get(key);
    if (!group) continue;
    if (change.property === "text-content") {
      group.textChange = {
        from: change.originalValue,
        to: change.newValue
      };
      continue;
    }
    const entry = {
      property: change.property,
      from: change.originalValue.trim() || "none",
      to: change.newValue,
      tailwind: cssToTailwind(change.property, change.newValue),
      tailwindClass: change.tailwindClass
    };
    group.changes.push(entry);
    if (TEXT_PROPERTIES2.has(change.property)) {
      group.childChanges.push(entry);
    }
  }
  return Array.from(groups.values());
}
function formatAllChanges(changes) {
  if (changes.length === 0) return "";
  const groups = groupChanges(changes);
  const route = getCurrentRoute();
  const lines = [];
  lines.push(`Make the following changes to the components on the \`${route}\` page:`);
  lines.push("");
  for (const group of groups) {
    const ancestry = getAncestryPath(group.selector);
    lines.push(`### [${group.tag}] ${group.text ? `"${group.text}"` : ""}`.trimEnd());
    lines.push(`  dom: ${ancestry || group.path}`);
    lines.push(`  classes: ${group.classes.join(" ")}`);
    if (group.textChange) {
      lines.push(`  Change text: "${group.textChange.from}" \u2192 "${group.textChange.to}"`);
    }
    if (group.changes.length > 0) {
      lines.push("  Apply these style updates:");
      for (const change of group.changes) {
        const displayValue = change.tailwindClass || change.tailwind;
        lines.push(`    ${change.property}: ${displayValue}`);
      }
    }
    if (group.childChanges.length > 0) {
      lines.push("  Also update child text elements (span, p, h1-h6, a, label):");
      for (const change of group.childChanges) {
        const displayValue = change.tailwindClass || change.tailwind;
        lines.push(`    ${change.property}: ${displayValue}`);
      }
    }
    lines.push("");
  }
  return lines.join("\n").trim();
}
async function copyToClipboard(text) {
  try {
    await navigator.clipboard.writeText(text);
    return true;
  } catch (e) {
    const textarea = document.createElement("textarea");
    textarea.value = text;
    textarea.style.position = "fixed";
    textarea.style.opacity = "0";
    document.body.appendChild(textarea);
    textarea.select();
    try {
      document.execCommand("copy");
      return true;
    } catch (e2) {
      return false;
    } finally {
      document.body.removeChild(textarea);
    }
  }
}
export {
  copyToClipboard,
  createInterfaceKitController,
  formatAllChanges
};
